### PR TITLE
Add whitelist regex support

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following:**

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
- [X] I have Signed Off all commits. ([`git commit --signoff`](https://git-scm.com/docs/git-commit#git-commit---signoff))

---

**What does this PR aim to accomplish?:**

Implement front end changes due to https://github.com/pi-hole/FTL/pull/612 and https://github.com/pi-hole/pi-hole/pull/2838

**How does this PR accomplish the above?:**

Generalize the already existing regex/whitelist handling we have for the blacklist to also cover the whitelist.

**What documentation changes (if any) are needed to support this PR?:**

Not much, as said, we only generalize features we already have now for the blacklist.

### CODE REVIEW

File: scripts/pi-hole/js/list.js [hunk 1]
```diff
@@ -43,12 +43,8 @@ function addListEntry(entry, index, list, button, type)
...
-    var listw;
     var list = $("#list");
-    if(listType === "black")
-    {
-        listw = $("#list-regex");
-    }
+    var listw = $("#list-regex");
     if(fade) {
         list.fadeOut(100);
         if(listw)
```
Comment (Mcat12) [inline scripts/pi-hole/js/list.js]: This `if` (and others) is no longer needed.

File: scripts/pi-hole/php/sub.php [hunk 2]
```diff
@@ -27,13 +27,16 @@
 
 switch($type) {
     case "white":
-        exec("sudo pihole -w -q -d ".$domain);
+        echo shell_exec("sudo pihole -w -q -d ".$domain);
         break;
     case "black":
-        exec("sudo pihole -b -q -d ".$domain);
+        echo shell_exec("sudo pihole -b -q -d ".$domain);
```
Comment (Mcat12) [inline scripts/pi-hole/php/sub.php]: What is the reason for this change?

File: scripts/pi-hole/php/auth.php [hunk 3]
```diff
@@ -11,7 +11,6 @@
 if (empty($ERRORLOG)) {
     $ERRORLOG = '/var/log/lighttpd/error.log';
 }
-$regexfile = "/etc/pihole/regex.list";
```
Comment (Mcat12) [inline scripts/pi-hole/php/auth.php]: The [`add_regex`](https://github.com/pi-hole/AdminLTE/blob/483a3aee27d12cc54109778aba8f0eff737d2c97/scripts/pi-hole/php/func.php#L48) function can also be removed now, as it probably isn't used by anything anymore.

File: scripts/pi-hole/php/add.php [hunk 4]
```diff
@@ -44,12 +44,18 @@
             echo shell_exec("sudo pihole -a audit ".$domains);
         }
         break;
-    case "regex":
+    case "black_regex":
         echo shell_exec("sudo pihole --regex --web ".$domains);
         break;
-    case "wild":
+    case "white_regex":
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: Can we add support for the whitelist regex to teleporter?

File: scripts/pi-hole/php/sub.php (hunk 2 above)
Comment (DL6ER) [reply scripts/pi-hole/php/sub.php]: This will make the `sub.php` page pass through warnings and errors. `add.php` does this as well.

File: scripts/pi-hole/php/add.php (hunk 4 above)
Comment (DL6ER) [reply scripts/pi-hole/php/add.php]: Done, I'll mark this PR as `[WIP]` as this new feature might need more testing.

**Comment by DL6ER** [comment]
If the warning
```
PHP Warning:  SQLite3Stmt::execute(): Unable to execute statement: unable to open database file in /var/www/html/admin/scripts/pi-hole/php/teleporter.php on line 131
```
shows up, this can be resolved via
```
sudo chmod g+w /etc/pihole/ /etc/pihole/gravity.db
sudo chown pihole:pihole /etc/pihole/gravity.db
sudo adduser www-data pihole
```
We should automate these steps.

File: scripts/pi-hole/php/teleporter.php [hunk 5]
```diff
@@ -29,68 +31,208 @@ function archive_add_file($path,$name,$subdir="")
...
+	{
+		$sql  = "INSERT OR IGNORE INTO ".$table;
+		$sql  .= " (id,domain,enabled,date_added,comment)";
+		$sql  .= " VALUES (:id,:domain,:enabled,:date_added,:comment);";
+	}
+
+	// Prepare SQLite statememt
+	$stmt = $db->prepare($sql);
+
+	// Return early if we prepare the SQLite statement
```
Comment (Mcat12) [inline scripts/pi-hole/php/teleporter.php]: Seems to be missing "fail to"

File: scripts/pi-hole/php/teleporter.php [hunk 6]
```diff
@@ -29,68 +31,208 @@ function archive_add_file($path,$name,$subdir="")
...
 
-function process_file($contents,$check=True)
+/**
+ * Create table rows from an uploaded archive file
+ *
+ * @param $file object The file of the file in the archive to import
+ * @param $table string The target table
+ * @param $flush boolean Whether to flush the table before importing the archived data
+ */
+function archive_insert_into_table($file, $table, $flush=false, $wildcardstyle=false)
```
Comment (Mcat12) [inline scripts/pi-hole/php/teleporter.php]: Documentation is missing `$wildcardstyle`

File: scripts/pi-hole/php/teleporter.php [hunk 7]
```diff
@@ -29,68 +31,208 @@ function archive_add_file($path,$name,$subdir="")
...
 {
-	$domains = array_filter(explode("\n",$contents));
+	global $db, $flushed_tables;
 
-	// Walk array and apply a max string length
-	// function to every member of the array of domains
-	array_walk($domains, "limit_length");
```
Comment (Mcat12) [inline scripts/pi-hole/php/teleporter.php]: This seems to be removed now. If I recall correctly, it prevents a DOS attack via injection of an expensive regex.

File: scripts/pi-hole/php/teleporter.php [hunk 8]
```diff
@@ -29,68 +31,208 @@ function archive_add_file($path,$name,$subdir="")
...
 
-function archive_add_directory($path,$subdir="")
+/**
+ * Restore the contents of a table from an uploaded archive
+ *
+ * @param $file object The file of the file in the archive to restore the table from
+ * @param $table string The table to import
+ * @param $flush boolean Whether to flush the table before importing the archived data
+ */
+function archive_restore_table($file, $table, $flush=false)
```
Comment (Mcat12) [inline scripts/pi-hole/php/teleporter.php]: These functions are missing the return documentation.

**Review by AzureMarker — COMMENTED** [review]
It may not be worthwhile to directly manipulate the database. Now there are issues with allowing PHP to write to the database, not to mention the extra complexity. It may be better to wait until the API to fully support the full table backups and restores.

File: scripts/pi-hole/php/teleporter.php (hunk 7 above)
Comment (DL6ER) [reply scripts/pi-hole/php/teleporter.php]: Should we re-add it? Is it meaningful to protect ourselves against an expensive regex by a mere length limitation?

File: scripts/pi-hole/php/add.php [hunk 9]
```diff
@@ -18,47 +18,90 @@
 
 // Don't check if the added item is a valid domain for regex expressions. Regex
 // filters are validated by FTL on import and skipped if invalid
-if($type !== "regex") {
+$check_lists = ["white","black","audit"];
+if(in_array($list, $check_lists)) {
     check_domain();
 }
 
 // Escape shell metacharacters
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: This comment is out of date

File: scripts/pi-hole/php/add.php [hunk 10]
```diff
@@ -18,47 +18,90 @@
...
-$domains = escapeshellcmd($_POST['domain']);
+$domains = explode(",",$_POST['domain']);
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: The domains are sent as whitespace separated values (exactly as the user entered them).

File: scripts/pi-hole/php/add.php [hunk 11]
```diff
@@ -18,47 +18,90 @@
...
-        break;
-    case "white_wild":
-        echo shell_exec("sudo pihole --whitewild --web ".$domains);
-        break;
-    case "audit":
-        echo shell_exec("sudo pihole -a audit ".$domains);
-        break;
+require_once("func.php");
+
+require("database.php");
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: Why are different require methods used?

File: scripts/pi-hole/php/add.php [hunk 12]
```diff
@@ -18,47 +18,90 @@
...
+{
+	global $db;
+	// Prepare SQLite statememt
+	$stmt = $db->prepare("INSERT OR IGNORE INTO ".$table." (domain) VALUES (:domain);");
+
+	// Return early if we prepare the SQLite statement
+	if(!$stmt)
+	{
+		echo "Failed to prepare statement for ".$table." table.";
+		echo $sql;
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: This variable does not exist.

File: scripts/pi-hole/php/add.php [hunk 13]
```diff
@@ -18,47 +18,90 @@
...
+		return 0;
+	}
+
+	// Loop over domains and inject the lines into the database
+	$num = 0;
+	foreach($domains as $row)
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: Use `$domain` instead of `$row` to be clear about what the variable represents.

File: scripts/pi-hole/php/add.php [hunk 14]
```diff
@@ -18,47 +18,90 @@
...
+	{
+		if($wildcardstyle)
+			$line = "(\\.|^)".str_replace(".","\\.",$row)."$";
+		else
+			$line = $row;
+
+		$stmt->bindValue(":domain", $line, SQLITE3_TEXT);
+
+		if($stmt->execute() && $stmt->reset() && $stmt->clear())
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: `clear` is not necessary as we will always re-bind the domain.

File: scripts/pi-hole/php/add.php [hunk 15]
```diff
@@ -18,47 +18,90 @@
...
+	global $db;
+	// Prepare SQLite statememt
+	$stmt = $db->prepare("INSERT OR IGNORE INTO ".$table." (domain) VALUES (:domain);");
+
+	// Return early if we prepare the SQLite statement
+	if(!$stmt)
+	{
+		echo "Failed to prepare statement for ".$table." table.";
+		echo $sql;
+		return 0;
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: This should return the error string like the rest of the function.

File: scripts/pi-hole/php/add.php [hunk 16]
```diff
@@ -18,47 +18,90 @@
...
-    case "audit":
-        echo shell_exec("sudo pihole -a audit ".$domains);
-        break;
+require_once("func.php");
+
+require("database.php");
+$GRAVITYDB = getGravityDBFilename();
+$db = SQLite3_connect($GRAVITYDB, SQLITE3_OPEN_READWRITE);
+
+function add_to_table($table, $domains, $wildcardstyle=false)
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: These functions should be moved to a separate file.

File: scripts/pi-hole/php/add.php [hunk 17]
```diff
@@ -18,47 +18,90 @@
...
+		if($stmt->execute() && $stmt->reset() && $stmt->clear())
+			$num++;
+		else
+		{
+			$stmt->close();
+			return "Error, added: ".$num."\n";
+		}
+	}
+
+	// Close database connection and return number or processed rows
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: Is "or" supposed to be "of?

File: scripts/pi-hole/php/teleporter.php (hunk 7 above)
Comment (Mcat12) [reply scripts/pi-hole/php/teleporter.php]: Here is the issue: #464. If we remove it without fixing it in a different way, the issue would have to be re-opened.

File: scripts/pi-hole/php/add.php [hunk 18]
```diff
@@ -8,100 +8,66 @@
...
+// Regex are validated by FTL during import
 $check_lists = ["white","black","audit"];
 if(in_array($list, $check_lists)) {
     check_domain();
 }
 
-// Escape shell metacharacters
-$domains = explode(",",$_POST['domain']);
+// Split individual domains into array
+$domains = explode(" ",$_POST['domain']);
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: Users may have multiple spaces between domains, which would cause this code to add empty domains to the list. Try this: https://stackoverflow.com/a/1792977

File: scripts/pi-hole/php/add.php [hunk 19]
```diff
@@ -8,100 +8,66 @@
...
 	case "audit":
-		echo add_to_table("domain_audit", $domains);
+		echo add_to_table($db, "domain_audit", $domains);
 		break;
+
+	default:
+		die("Invalid list!");
 }
+
+// Reload lists in pihole-FTL after having removed something
```
Comment (Mcat12) [inline scripts/pi-hole/php/add.php]: This comment should say "added" instead of "removed"

File: scripts/pi-hole/php/database.php [hunk 20]
```diff
@@ -54,4 +54,90 @@ function SQLite3_connect($filename, $mode=SQLITE3_OPEN_READONLY)
...
+
+/**
+ * Add domains to a given table
+ *
+ * @param $db object The SQLite3 database connection object
+ * @param $table string The target table
+ * @param $domains array Array of domains (strings) to be added to the table
+ * @param $wildcardstyle boolean Whether to format the input domains in legacy wildcard notation
+ * @return string Success/error and number of processed domains
+ */
```
Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: Thanks for adding docs!

File: scripts/pi-hole/php/database.php [hunk 21]
```diff
@@ -54,4 +54,90 @@ function SQLite3_connect($filename, $mode=SQLITE3_OPEN_READONLY)
...
+function add_to_table($db, $table, $domains, $wildcardstyle=false)
+{
+	// Prepare SQLite statememt
+	$stmt = $db->prepare("INSERT OR IGNORE INTO ".$table." (domain) VALUES (:domain);");
+
+	// Return early if we prepare the SQLite statement
+	if(!$stmt)
+	{
+		echo "Failed to prepare statement for ".$table." table.";
+		return "Error, added: 0\n";
```
Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: Is the newline at the end necessary?

Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: Are the newlines at the end necessary?

File: scripts/pi-hole/php/database.php [hunk 22]
```diff
@@ -54,4 +54,90 @@ function SQLite3_connect($filename, $mode=SQLITE3_OPEN_READONLY)
...
+ * @param $domains array Array of domains (strings) to be added to the table
+ * @param $wildcardstyle boolean Whether to format the input domains in legacy wildcard notation
+ * @return string Success/error and number of processed domains
+ */
+function add_to_table($db, $table, $domains, $wildcardstyle=false)
+{
+	// Prepare SQLite statememt
+	$stmt = $db->prepare("INSERT OR IGNORE INTO ".$table." (domain) VALUES (:domain);");
+
+	// Return early if we prepare the SQLite statement
```
Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: Seems to be missing "fail to"

File: scripts/pi-hole/php/database.php [hunk 23]
```diff
@@ -54,4 +54,90 @@ function SQLite3_connect($filename, $mode=SQLITE3_OPEN_READONLY)
...
+		if($stmt->execute() && $stmt->reset())
+			$num++;
+		else
+		{
+			$stmt->close();
+			return "Error, added: ".$num."\n";
+		}
+	}
+
+	// Close database connection and return number of processed rows
```
Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: This is closing the prepared statement, not the DB connection.

File: scripts/pi-hole/php/teleporter.php [hunk 24]
```diff
@@ -205,6 +210,10 @@ function archive_insert_into_table($file, $table, $flush=false, $wildcardstyle=f
 		else
```
Comment (Mcat12) [inline scripts/pi-hole/php/teleporter.php]: This code is very similar to the other add/sub functions. Can this code use those functions instead?

File: scripts/pi-hole/php/database.php (hunk 21 above)
Comment (DL6ER) [reply scripts/pi-hole/php/database.php]: When we black-/whitelist from the audit page, we add to both, the respective list *and* the audit list at the same time. Without this newline, the outputs from the two calls would appear on the same line.

Comment (Mcat12) [reply scripts/pi-hole/php/database.php]: The front-end should handles this instead.

File: scripts/pi-hole/php/teleporter.php (hunk 24 above)
Comment (DL6ER) [reply scripts/pi-hole/php/teleporter.php]: These routines restore from a file instead of receiving an array. We can write a wrapper that calls `add_to_table()` for the legacy import (where we only import the domains).

Comment (Mcat12) [reply scripts/pi-hole/php/teleporter.php]: It does load the domains into an array from the file, so at some point the code does become similar enough. The file loading and table flushing code can be kept in this file, but the insertion can be done through the separate method.

File: scripts/pi-hole/php/database.php [hunk 25]
```diff
@@ -81,6 +85,10 @@ function add_to_table($db, $table, $domains, $wildcardstyle=false)
 	$num = 0;
 	foreach($domains as $domain)
 	{
+		// Skip empty domains
+		if(strlen($domain) == 0)
+			continue;
```
Comment (Mcat12) [inline scripts/pi-hole/php/database.php]: This should be an impossible case now that the input domain list is properly split. It can be assumed that the input domain list contains valid data.

File: scripts/pi-hole/php/teleporter.php (hunk 7 above)
Comment (DL6ER) [reply scripts/pi-hole/php/teleporter.php]: This is handled by

https://github.com/pi-hole/AdminLTE/pull/986/files#diff-3e979b46422dde1d81341526e434d794R123

I will add the same code in `add_to_table()`.

File: settings.php [hunk 26]
```diff
@@ -1155,7 +1160,12 @@ function convertseconds($argument)
                                                     <div class="checkbox">
                                                         <label><input type="checkbox" name="regexlist" value="true"
                                                                       checked>
-                                                            Regex filters</label>
+                                                            Blacklist (regex)</label>
+                                                    </div>
+                                                    <div class="checkbox">
+                                                        <label><input type="checkbox" name="adlist" value="true"
+                                                                      checked>
+                                                            Lists of Ads</label>
```
Comment (Mcat12) [inline settings.php]: This should be "Blocklists"

File: settings.php [hunk 27]
```diff
@@ -1172,6 +1182,11 @@ function convertseconds($argument)
                                                     <button type="submit" class="btn btn-default" name="action"
                                                             value="in">Import
                                                     </button>
+                                                    <div class="checkbox">
+                                                        <label><input type="checkbox" name="flushtables" value="true"
+                                                                      checked>
+                                                            Flush tables during import</label>
```
Comment (Mcat12) [inline settings.php]: This may be too technical, perhaps instead "Clear existing data".

**Comment by Mcat12** [comment]
Clicking enter to add an entry to a list results in "Invalid list!", due to the code using `exact` as the list type.

There is no error when adding a domain which already exists. It just does nothing and returns with success.

**Comment by DL6ER** [comment]
SQLite3 errors are now passed through to the user:
Screenshot at 2019-08-27 18-53-44

**Comment by DL6ER** [comment]
> There is no error when adding a domain which already exists. It just does nothing and returns with success.

Yes, this is expected and part of the conflict handling (`INSERT OR IGNORE`). As we use the same routine in the teleporter, I wouldn't like to change it to return an error on already existing domains.

I changed the code to specify how many domains (out of how many) have actually been added to the database.
Screenshot at 2019-08-27 19-28-18

After hitting return:
Screenshot at 2019-08-27 19-28-33

**Comment by DL6ER** [comment]
I updated the message
Screenshot at 2019-08-29 16-46-24

Screenshot at 2019-08-29 16-47-26

**Review by AzureMarker — APPROVED** [review]
